crates/core/tedge_api/src/mqtt_topics.rs

```diff
@@ -0,0 +1,407 @@
+//! This module abstracts the MQTT topics used by thin-edge.
+//!
+//! See https://thin-edge.github.io/thin-edge.io/next/references/mqtt-api/
+
+use std::str::FromStr;
+
+/// The MQTT topics are represented by three distinct groups:
+/// - a root prefix, used by all the topics
+/// - an entity topic identifier of the source or target of the messages
+/// - a channel kind for the messages exchanged along this topic
+///
+/// Once built from a root prefix, the main two features of such a schema are to:
+/// - get the topic addressing a given entity channel
+/// - get the entity channel addressed by some topic
+///
+/// ```
+/// # use tedge_api::mqtt_topics::{MqttSchema, Channel, EntityTopicId};
+/// # use mqtt_channel::Topic;
+///
+/// // The default root prefix is `"te"`:
+/// let te = MqttSchema::default();
+/// assert_eq!(&te.root, "te");
+///
+/// // Getting the entity channel addressed by some topic
+/// let topic = Topic::new_unchecked("te/device/child001/service/service001/m/measurement_type");
+/// let entity: EntityTopicId = "device/child001/service/service001".parse().unwrap();
+/// let channel = Channel::Measurement {
+///     measurement_type: "measurement_type".to_string(),
+/// };
+/// assert_eq!(
+///     te.entity_channel_of(&topic).ok(),
+///     Some((entity.clone(), channel.clone()))
+/// );
+///
+/// // Getting the topic to address a specific entity channel
+/// assert_eq!(
+///     te.topic_for(&entity, &channel).name,
+///     topic.name
+/// );
+/// ```
+pub struct MqttSchema {
+    pub root: String,
+}
+
+/// The default root prefix used by thin-edge is `te`
+impl Default for MqttSchema {
+    fn default() -> Self {
+        MqttSchema::new("te")
+    }
+}
+
+impl MqttSchema {
+    /// Build a new schema using the given root prefix for all topics.
+    pub fn new(root: &str) -> Self {
+        MqttSchema {
+            root: root.to_string(),
+        }
+    }
+
+    /// Get the topic addressing a given entity channel
+    pub fn topic_for(&self, entity: &EntityTopicId, channel: &Channel) -> mqtt_channel::Topic {
+        let topic = format!("{}/{}/{}", self.root, entity.0, channel.to_string());
+        mqtt_channel::Topic::new(&topic).unwrap()
+    }
+
+    /// Get the entity channel addressed by some topic
+    pub fn entity_channel_of(
+        &self,
+        topic: &mqtt_channel::Topic,
+    ) -> Result<(EntityTopicId, Channel), EntityTopicError> {
+        self.parse(&topic.name)
+    }
+}
+
+impl MqttSchema {
+    fn parse(&self, topic: &str) -> Result<(EntityTopicId, Channel), EntityTopicError> {
```

Comment:
Indeed, this can be discussed.

I decided to remove `EntityTopic` in favor of a tuple, after all the methods previously attached to  `EntityTopic` have been moved either to `EntityTopicId` or `Channel`.

Reply:
> I decided to remove `EntityTopic` in favor of a tuple, after all the methods previously attached to `EntityTopic` have been moved either to `EntityTopicId` or `Channel`.

Moving methods that operate on topic identifiers and channels to their own respective types is all well and good, I just think that splitting a topic into an identifier and a channel, just to pass them as a separate arguments into the same function unnecessarily adds arguments. Also when taking arguments separately, they don't have to even be related and one could use completely separate topic id and channel, which is maybe a bit nitpicky, so the final call is yours. From my side though it would be good to have `EntityTopic`, even if it's just a minimal container for `EntityTopicId` and `Channel`.

Reply:
> From my side though it would be good to have `EntityTopic`, even if it's just a minimal container for `EntityTopicId` and `Channel`.

What I propose, is to consider re-introducing  `EntityTopic` during the on-going work refactoring the c8y mapper to use the entity-store and the mqtt-schema. During this work, we will have a better understanding of whether or not having such a structure is a good idea.